ops/framework.py

```diff
@@ -88,7 +88,48 @@ def from_path(cls, path):
         return handle
 
 
-class EventBase:
+class UnassignedEvent(type):
+    """Metaclass for events which have not yet been assigned to an emitter type.
+
+    Without being assigned and bound to an emitter, unassigned events cannot be emitted.
+    """
+
+    def __set_name__(event_type, emitter_type, event_kind):
+        # Permanently assign the event for this emitter type.
+        setattr(emitter_type, event_kind, AssignedEvent(event_type, event_kind, emitter_type))
```

Comment:
This is the actual meat of this PR.  This line replacing `Event(SomeEvent)` by deep Python magic (metaclasses, special methods, etc) that doesn't make sense until we dig down. I'm usually on the fence about these things, as it'll obscure code that people will be typing themselves.. this:
```
class C:
     foo = Bar
```
doesn't do what it says on the tin anymore. Literally, Class.foo is not Bar.

With that said, we can certainly give it a shot and see how people feel in practice. But this PR is doing a lot more than that, which is not great. What pieces of this PR are indeed required to achieve this goal, vs. things which are irrelevant for this change?

How about this: `AssignedEvent` is really our current `Event` type. I think we can be precise and name it as `EventDescriptor` now that we're not typing it all out anymore. But let's not do that now.. let's keep it named `Event`, and try not to touch it. With that, and this metaclass, what else do we need on this PR for this change to work out?

Reply:
I definitely see the point that this replaces the explicit wrapping of the decorator with implicit behavior that might be surprising, but I think that it ends up being worth it when weighed against the possibility that a new charm author might not realize they need to always wrap the event type with the `Event` descriptor, or even that an experienced charm author might forget.

At the end of the day, I think the charm author really will care about defining an event and having `self.on.foo.emit()` work and not so much about how that ends up working under the hood.  There is also a parallel to how instance methods work in Python; like instance methods, having events become bound is the common case and thus should be the default, just like there is no `@instancemethod` decorator required in Python and the binding just happens automatically because that's almost always what you intend.